gnupg-pkcs11-scd/keyutil.c

```diff
@@ -118,43 +217,87 @@ keyutil_get_cert_mpi (
 		error = GPG_ERR_BAD_CERT;
 		goto cleanup;
 	}
- 
+
 	if ((pubkey = X509_get_pubkey (x509)) == NULL) {
 		error = GPG_ERR_BAD_CERT;
 		goto cleanup;
 	}
- 
-	if ((rsa = EVP_PKEY_get1_RSA(pubkey)) == NULL) {
-		error = GPG_ERR_WRONG_PUBKEY_ALGO;
+
+	pubkey_ctx = EVP_PKEY_CTX_new(pubkey, NULL);
+	if (pubkey_ctx == NULL) {
+		error = GPG_ERR_BAD_CERT;
 		goto cleanup;
 	}
 
-	RSA_get0_key(rsa, &n, &e, NULL);
-
-	n_hex = BN_bn2hex (n);
-	e_hex = BN_bn2hex (e);
-
-	if(n_hex == NULL || e_hex == NULL) {
-		error = GPG_ERR_BAD_KEY;
+	/**
+	 * Check the public key context
+	 * 1 is success, -2 is not applicable
+	 */
+	check_result = EVP_PKEY_public_check(pubkey_ctx);
+	if (check_result != 1 && check_result != -2) {
+		error = GPG_ERR_WRONG_PUBKEY_ALGO;
 		goto cleanup;
 	}
- 
-	if (
-		gcry_mpi_scan (&n_mpi, GCRYMPI_FMT_HEX, n_hex, 0, NULL) ||
-		gcry_mpi_scan (&e_mpi, GCRYMPI_FMT_HEX, e_hex, 0, NULL)
-	) {
-		error = GPG_ERR_BAD_KEY;
-		goto cleanup;
+
+	if (EVP_PKEY_base_id(pubkey) == EVP_PKEY_RSA) {
+		keyinfo_init(keyinfo, KEYINFO_KEY_TYPE_RSA);
+	}
+
+	switch (keyinfo->type) {
+		case KEYINFO_KEY_TYPE_RSA:
+			/* Warning: EVP_PKEY_get1_RSA is deprecated in OpenSSL 3.0 */
+			if ((rsa = EVP_PKEY_get1_RSA(pubkey)) == NULL) {
+				error = GPG_ERR_WRONG_PUBKEY_ALGO;
+				goto cleanup;
+			}
+
+			/* Warning: RSA_get0_key is deprecated in OpenSSL 3.0 */
+			RSA_get0_key(rsa, &n, &e, NULL);
+
+			/* Warning: RSA_size is deprecated in OpenSSL 3.0 */
+			keyinfo->key_length = RSA_size(rsa) * 8;
+
+			n_hex = BN_bn2hex (n);
+			e_hex = BN_bn2hex (e);
+
+			if(n_hex == NULL || e_hex == NULL) {
+				error = GPG_ERR_BAD_KEY;
+				goto cleanup;
+			}
+
+			if (
+				gcry_mpi_scan (&n_mpi, GCRYMPI_FMT_HEX, n_hex, 0, NULL) ||
+				gcry_mpi_scan (&e_mpi, GCRYMPI_FMT_HEX, e_hex, 0, NULL)
+			) {
+				error = GPG_ERR_BAD_KEY;
+				goto cleanup;
+			}
+			break;
+		case KEYINFO_KEY_TYPE_UNKNOWN:
+		case KEYINFO_KEY_TYPE_INVALID:
+			error = GPG_ERR_BAD_KEY;
+			goto cleanup;
 	}
 #else
 #error Invalid configuration.
 #endif
 
-	*p_n_mpi = n_mpi;
-	n_mpi = NULL;
-	*p_e_mpi = e_mpi;
-	e_mpi = NULL;
-	error = GPG_ERR_NO_ERROR;
+	switch (keyinfo->type) {
```

Comment:
why do we switch again for same variable?

we should put the `*_mpi` directly in the `keyinfo`, it is ok to partially initialize it and return an error to simplify the code.

Reply:
The idea is that GNUTLS and OpenSSL will both need to do their own specific kinds of work to support computing these values from the X.509 certificate and so this case represents the common code.

Reply:
I do not think that this common code is required if you accept partial initialization of keyinfo when error.